File: PythonCodes/Fig10_02.py
#%%
"""
Created on Feb 11 2019
Local volatility model based on the implied volatility obtained from prices from the Heston model
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import scipy.optimize as optimize
import enum 

# This class defines puts and calls

class OptionType(enum.Enum):
    CALL = 1.0
    PUT = -1.0
    
def CallPutOptionPriceCOSMthd(cf,CP,S0,r,tau,K,N,L):

    # cf   - Characteristic function, in the book denoted by \varphi
    # CP   - C for call and P for put
    # S0   - Initial stock price
    # r    - Interest rate (constant)
    # tau  - Time to maturity
    # K    - List of strikes
    # N    - Number of expansion terms
    # L    - Size of truncation domain (typ.:L=8 or L=10)  
        
    # Reshape K to become a column vector

    if K is not np.array:
        K = np.array(K).reshape([len(K),1])
    
    # Assigning i=sqrt(-1)

    i = np.complex(0.0,1.0) 
    x0 = np.log(S0 / K)   
    
    # Truncation domain

    a = 0.0 - L * np.sqrt(tau)
    b = 0.0 + L * np.sqrt(tau)
    
    # Summation from k = 0 to k=N-1

    k = np.linspace(0,N-1,N).reshape([N,1])  
    u = k * np.pi / (b - a);  

    # Determine coefficients for put prices  

    H_k = CallPutCoefficients(CP,a,b,k)   
    mat = np.exp(i * np.outer((x0 - a) , u))
    temp = cf(u) * H_k 
    temp[0] = 0.5 * temp[0]    
    value = np.exp(-r * tau) * K * np.real(mat.dot(temp))     
    return value

# Determine coefficients for put prices 

def CallPutCoefficients(CP,a,b,k):
    if CP==OptionType.CALL:                  
        c = 0.0
        d = b
        coef = Chi_Psi(a,b,c,d,k)
        Chi_k = coef["chi"]
        Psi_k = coef["psi"]
        if a < b and b < 0.0:
            H_k = np.zeros([len(k),1])
        else:
            H_k      = 2.0 / (b - a) * (Chi_k - Psi_k)  
    elif CP==OptionType.PUT:
        c = a
        d = 0.0
        coef = Chi_Psi(a,b,c,d,k)
        Chi_k = coef["chi"]
        Psi_k = coef["psi"]
        H_k      = 2.0 / (b - a) * (- Chi_k + Psi_k)               
    
    return H_k    

def Chi_Psi(a,b,c,d,k):
    psi = np.sin(k * np.pi * (d - a) / (b - a)) - np.sin(k * np.pi * (c - a)/(b - a))
    psi[1:] = psi[1:] * (b - a) / (k[1:] * np.pi)
    psi[0] = d - c
    
    chi = 1.0 / (1.0 + np.power((k * np.pi / (b - a)) , 2.0)) 
    expr1 = np.cos(k * np.pi * (d - a)/(b - a)) * np.exp(d)  - np.cos(k * np.pi 
                  * (c - a) / (b - a)) * np.exp(c)
    expr2 = k * np.pi / (b - a) * np.sin(k * np.pi * 
                        (d - a) / (b - a))   - k * np.pi / (b - a) * np.sin(k 
                        * np.pi * (c - a) / (b - a)) * np.exp(c)
    chi = chi * (expr1 + expr2)
    
    value = {"chi":chi,"psi":psi }
    return value
 
    # Black-Scholes call option price

def BS_Call_Put_Option_Price(CP,S_0,K,sigma,tau,r):
    if K is list:
        K = np.array(K).reshape([len(K),1])
    d1    = (np.log(S_0 / K) + (r + 0.5 * np.power(sigma,2.0)) 
    * tau) / (sigma * np.sqrt(tau))
    d2    = d1 - sigma * np.sqrt(tau)
    if CP == OptionType.CALL:
        value = st.norm.cdf(d1) * S_0 - st.norm.cdf(d2) * K * np.exp(-r * tau)
    elif CP == OptionType.PUT:
        value = st.norm.cdf(-d2) * K * np.exp(-r * tau) - st.norm.cdf(-d1)*S_0
    return value

# Implied volatility method

def ImpliedVolatility(CP,marketPrice,K,T,S_0,r):

    # To determine initial volatility we define a grid for sigma
    # and interpolate on the inverse function

    sigmaGrid = np.linspace(0,2,200)
    optPriceGrid = BS_Call_Put_Option_Price(CP,S_0,K,sigmaGrid,T,r)
    sigmaInitial = np.interp(marketPrice,optPriceGrid,sigmaGrid)
    print("Initial volatility = {0}".format(sigmaInitial))
    
    # Use already determined input for the local-search (final tuning)

    func = lambda sigma: np.power(BS_Call_Put_Option_Price(CP,S_0,K,sigma,T,r) - marketPrice, 1.0)
    impliedVol = optimize.newton(func, sigmaInitial, tol=1e-15)
    print("Final volatility = {0}".format(impliedVol))
    return impliedVol

def ChFHestonModel(r,tau,kappa,gamma,vbar,v0,rho):
    i = np.complex(0.0,1.0)
    D1 = lambda u: np.sqrt(np.power(kappa-gamma*rho*i*u,2)+(u*u+i*u)*gamma*gamma)
    g  = lambda u: (kappa-gamma*rho*i*u-D1(u))/(kappa-gamma*rho*i*u+D1(u))
    C  = lambda u: (1.0-np.exp(-D1(u)*tau))/(gamma*gamma*(1.0-g(u)*np.exp(-D1(u)*tau)))\
        *(kappa-gamma*rho*i*u-D1(u))

    # Note that we exclude the term -r*tau, as the discounting is performed in the COS method

    A  = lambda u: r * i*u *tau + kappa*vbar*tau/gamma/gamma *(kappa-gamma*rho*i*u-D1(u))\
        - 2*kappa*vbar/gamma/gamma*np.log((1.0-g(u)*np.exp(-D1(u)*tau))/(1.0-g(u)))

    # Characteristic function for the Heston model    

    cf = lambda u: np.exp(A(u) + C(u)*v0)
    return cf 

def OptionPriceWithCosMethodHelp(CP,T,K,S0,r,kappa,gamma,v0,vbar,rho):

     # Settings for the COS method

    N = 500
    L = 8
           
    # ChF for the Heston model

    cf = ChFHestonModel(r,T,kappa,gamma,vbar,v0,rho)
    OptPrice = CallPutOptionPriceCOSMthd(cf, CP, S0, r, T, K, N, L)   
    return OptPrice
    
def EUOptionPriceFromMCPaths(CP,S,K,T,r):

    # S is a vector of Monte Carlo samples at T

    if CP == OptionType.CALL:
        return np.exp(-r*T)*np.mean(np.maximum(S-K,0.0))
    elif CP == OptionType.PUT:
        return np.exp(-r*T)*np.mean(np.maximum(K-S,0.0))
    
def mainCalculation():

    # Heston model parameters   

    kappa   = 1.3
    vbar    = 0.05
    gamma   = 0.3 
    rho     = -0.3 
    v0      = 0.1
    r       = 0.06    
    S0     = 1.0
    T      = 1.0
    r      = 0.06
    CP     = OptionType.CALL

    # Monte Carlo settings

    NoOfPaths = 5000
    NoOfSteps = (int)(250*T)
    
       
    # Define a function to calculate option prices

    V =lambda T,K: OptionPriceWithCosMethodHelp(CP,T,K,S0,r,kappa,gamma,v0,vbar,rho)
   
    # Define a shock size for derivative calculation

    bump_T  = 1e-4
    bump_K  = 1e-4
    
    # Define derivatives

    dV_dT   = lambda T,K: (V(T + bump_T,K) - V(T ,K)) / bump_T
    dV_dK   = lambda T,K: (V(T,K + bump_K) - V(T,K - bump_K)) / (2.0 * bump_K)
    d2V_dK2 = lambda T,K: (V(T,K + bump_K) + V(T,K-bump_K) - 2.0*V(T,K))/(bump_K**2.0 )

    # Local variance

    sigmaLV2=lambda T,K: (dV_dT(T,K) + r * K * dV_dK(T,K)) / (0.5 * K**2.0 * d2V_dK2(T,K))

    # Monte Carlo simulation

    dt        = T/NoOfSteps
    np.random.seed(5)
    Z         = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    S         = np.zeros([NoOfPaths,NoOfSteps+1])
    
    S[:,0]    = S0;
    time = np.zeros([NoOfSteps+1,1])
    
    for i in range(0,NoOfSteps):

        # This condition is necessary as for t=0 we cannot compute implied
        # volatilities

        if time[i]==0.0:
            time[i]=0.0001
        
        print('current time is {0}'.format(time[i]))

        # Standarize Normal(0,1)

        Z[:,i]=(Z[:,i]-np.mean(Z[:,i]))/np.std(Z[:,i])
        
        # Compute local volatility

        S_i = np.array(S[:,i]).reshape([NoOfPaths,1])
        
        if time[i]  >0 :
            sig = np.real(sigmaLV2(time[i],S_i))
        elif time[i]==0:
            sig = np.real(sigmaLV2(dt/2.0,S_i))
        
        np.nan_to_num(sig)
        
        # Because of discretizations we may encouter negative variance which
        # are set to 0 here.

        sig=np.maximum(sig,1e-14)
        sigmaLV = np.sqrt(sig.transpose())  
        
        # Stock path

        S[:,i+1] = S[:,i] * (1.0 + r*dt + sigmaLV*Z[:,i]*np.sqrt(dt))
                       
        # We force that at each time S(t)/M(t) is a martingale

        S[:,i+1] = S[:,i+1] - np.mean(S[:,i+1]) + S0*np.exp(r*(time[i]+dt))
        
        # Make sure that after moment matching we don't encounter negative stock values

        S[:,i+1] = np.maximum(S[:,i+1],1e-14)
        
        # Adjust time

        time[i+1] = time[i] + dt
        
    # Plot some results

    K = np.linspace(0.5,1.7,50)
    OptPrice = np.zeros([len(K),1])
    IV_Heston = np.zeros([len(K),1])
    IV_MC = np.zeros([len(K),1])
    
    # Prices from the Heston model

    valCOS = V(T,K)
    
    # Implied volatilities

    for (idx,k) in enumerate(K):
        OptPrice[idx] = EUOptionPriceFromMCPaths(CP,S[:,-1],k,T,r)
        IV_MC[idx]    = ImpliedVolatility(CP,OptPrice[idx],k,T,S0,r)*100.0
        IV_Heston[idx] =ImpliedVolatility(CP,valCOS[idx],K[idx],T,S0,r)*100
        
    # Plot the option prices

    plt.figure(1)
    plt.plot(K,OptPrice)
    plt.grid()
    plt.xlabel('strike')
    plt.ylabel('option price')
    
    # Plot the implied volatilities

    plt.figure(2)
    plt.plot(K,IV_Heston)
    plt.plot(K,IV_MC,'-r')
    plt.grid()
    plt.xlabel('strike')
    plt.ylabel('implied volatility')
    plt.legend(['Heston','Monte Carlo'])
    plt.axis([np.min(K),np.max(K),0,40])
    plt.ylim([25,35])
mainCalculation()

File: MatlabCodes/Chapter 10/Fig10_02.m
function LocalVolHeston
format long;close all;
randn('seed',8); % Random seed
NoOfPaths  =10000;

CP         ='c';
S0         =1;       % Initial stock
r          =0.06;    % Interest rate
tau        =1;       % Maturity
NoOfSteps  =100*tau;
dt         =tau/NoOfSteps; %Tme increament

% Heston model parameters

kappa   = 1.3; 
vbar    = 0.05; 
gamma   = 0.3; 
rho     = -0.3; 
v0      = 0.1;

% Define a function to calculate option prices

V   = @(T,K)OptionPriceWithCosMethodHelp(CP,T,K,S0,r,kappa,gamma,v0,vbar,rho);

% Define bump size

bump_T  = 1e-4;
bump_K  = @(T)1e-4;

% Define derivatives

dC_dT   = @(T,K) (V(T + bump_T,K) - V(T ,K)) /  bump_T;
dC_dK   = @(T,K) (V(T,K + bump_K(T)) - V(T,K - bump_K(T))) / (2 * bump_K(T));
d2C_dK2 = @(T,K) (V(T,K + bump_K(T)) + V(T,K-bump_K(T)) - 2*V(T,K)) / bump_K(T)^2;

% Local volatility

sigma2LV=@(T,K)(dC_dT(T,K) + r * K .* dC_dK(T,K))...
        ./ (0.5 .* K.^2 .* d2C_dK2(T,K));

% Prepare stock and Brownian motion

S       =zeros(NoOfPaths,1);
S(:,1)  =S0;
Z       =random('normal',0,sqrt(dt),[NoOfPaths,NoOfSteps]);

t=0;
for i=1:NoOfSteps

    % Adjust noise so that ~ N(0,sqrt(dt))

    Z(:,i)=(Z(:,i)-mean(Z(:,i)))/std(Z(:,i))*sqrt(dt);
    if t==0
                sigma_t =sqrt(sigma2LV(0.001, S));
    else
                sigma_t =sqrt(max(sigma2LV(t, S),0));         
    end
    S = S .* (1 + r * dt + sigma_t .* Z(:,i));
    
    S=(S-mean(S))+S0*exp(r*(t+dt));
    S= max(S,0);
    t = t + dt                         
end

% Calculate implied volatilities: local vol. vs. the Heston model

K          =0.5:0.1:1.7; % Strikes at which smile is determined

% Initialization of all variables (increases Matlab efficiency)

error      =zeros(length(K),1);
call_LV=error; imp_vol_LV=error; imp_vol_Heston=error;

% For each strike, determine the implied volatility

OptionValueHeston    = V(tau,K');
for j=1:length(K)

    % Local volatility

    call_LV(j)        = exp(-r * tau) * mean( max(S - K(j),0) );
    imp_vol_LV(j)     = ImpliedVolatility(CP,call_LV(j),K(j),tau,S0,r,0.2);

    % Heston

    imp_vol_Heston(j) = ImpliedVolatility(CP,OptionValueHeston(j),K(j),tau,S0,r,0.2);    

    % Error

    error(j)=imp_vol_LV(j)-imp_vol_Heston(j);  
end
figure(1)
hold on
plot(K,imp_vol_LV*100,'linewidth',1.5)
plot(K,imp_vol_Heston*100,'--r','linewidth',1.5)
xlabel('strike, K')
ylabel('Implied volatility [%]')
grid on
legend('Local volatility','Stochastic Volatility (Heston)')
xlim([min(K),max(K)])

function value = OptionPriceWithCosMethodHelp(CP,T,K,S0,r,kappa,gamma,v0,vbar,rho)
cf   = @(u) ChFHeston(u, T, kappa,vbar,gamma,rho, v0, r);

% The COS method

% Settings for the COS method

N = 500;
L = 8;
value = CallPutOptionPriceCOSMthd(cf,CP,S0,r,T,K,N,L);

function cf=ChFHeston(u, tau, kappa,vBar,gamma,rho, v0, r)
i     = complex(0,1);

% Functions D_1 and g

D_1  = sqrt(((kappa -i*rho*gamma.*u).^2+(u.^2+i*u)*gamma^2));
g    = (kappa- i*rho*gamma*u-D_1)./(kappa-i*rho*gamma*u+D_1);    

% Complex-valued functions A and C

C = (1/gamma^2)*(1-exp(-D_1*tau))./(1-g.*exp(-D_1*tau)).*(kappa-gamma*rho*i*u-D_1);
A = i*u*r*tau + kappa*vBar*tau/gamma^2 * (kappa-gamma*rho*i*u-D_1)-2*kappa*vBar/gamma^2*log((1-g.*exp(-D_1*tau))./(1-g));

% ChF for the Heston model

cf = exp(A + C * v0);

function value = CallPutOptionPriceCOSMthd(cf,CP,S0,r,tau,K,N,L)
i = complex(0,1);


% cf   - Characteristic function, in the book denoted as \varphi
% CP   - C for call and P for put
% S0   - Initial stock price
% r    - Interest rate (constant)
% tau  - Time to maturity
% K    - Vector of strike prices
% N    - Number of expansion terms
% L    - Size of truncation domain (typ.:L=8 or L=10)

x0 = log(S0 ./ K);   

% Truncation domain

a = 0 - L * sqrt(tau); 
b = 0 + L * sqrt(tau);

k = 0:N-1;              % Row vector, index for expansion terms
u = k * pi / (b - a);   % ChF arguments

H_k = CallPutCoefficients(CP,a,b,k);
temp    = (cf(u) .* H_k).';
temp(1) = 0.5 * temp(1);      % Multiply the first element by 1/2

mat = exp(i * (x0 - a) * u);  % Matrix-vector manipulations

% Final output

value = exp(-r * tau) * K .* real(mat * temp);

% Coefficients H_k for the COS method

function H_k = CallPutCoefficients(CP,a,b,k)
    if lower(CP) == 'c' || CP == 1
        c = 0;
        d = b;
        [Chi_k,Psi_k] = Chi_Psi(a,b,c,d,k);
         if a < b && b < 0.0
            H_k = zeros([length(k),1]);
         else
            H_k = 2.0 / (b - a) * (Chi_k - Psi_k);
         end
    elseif lower(CP) == 'p' || CP == -1
        c = a;
        d = 0.0;
        [Chi_k,Psi_k]  = Chi_Psi(a,b,c,d,k);
         H_k = 2.0 / (b - a) * (- Chi_k + Psi_k);       
    end

function [chi_k,psi_k] = Chi_Psi(a,b,c,d,k)
    psi_k        = sin(k * pi * (d - a) / (b - a)) - sin(k * pi * (c - a)/(b - a));
    psi_k(2:end) = psi_k(2:end) * (b - a) ./ (k(2:end) * pi);
    psi_k(1)     = d - c;
    
    chi_k = 1.0 ./ (1.0 + (k * pi / (b - a)).^2); 
    expr1 = cos(k * pi * (d - a)/(b - a)) * exp(d)  - cos(k * pi... 
                  * (c - a) / (b - a)) * exp(c);
    expr2 = k * pi / (b - a) .* sin(k * pi * ...
                        (d - a) / (b - a))   - k * pi / (b - a) .* sin(k... 
                        * pi * (c - a) / (b - a)) * exp(c);
    chi_k = chi_k .* (expr1 + expr2);
    
% Closed-form expression of European call/put option with Black-Scholes formula

function value=BS_Call_Option_Price(CP,S_0,K,sigma,tau,r)

% Black-Scholes call option price

d1    = (log(S_0 ./ K) + (r + 0.5 * sigma^2) * tau) / (sigma * sqrt(tau));
d2    = d1 - sigma * sqrt(tau);
if lower(CP) == 'c' || lower(CP) == 1
    value =normcdf(d1) * S_0 - normcdf(d2) .* K * exp(-r * tau);
elseif lower(CP) == 'p' || lower(CP) == -1
    value =normcdf(-d2) .* K*exp(-r*tau) - normcdf(-d1)*S_0;
end

function impliedVol = ImpliedVolatility(CP,marketPrice,K,T,S_0,r,initialVol)
    func = @(sigma) (BS_Call_Option_Price(CP,S_0,K,sigma,T,r) - marketPrice).^1.0;
    impliedVol = fzero(func,initialVol);
